Python 3.13: test_assert_valid_name_minimal fails due to changed behavior of PurePath.is_reserved()

According to [upstream's documentation](https://docs.python.org/3.13/library/pathlib.html#pathlib.PurePath.is_reserved) the behavior of `PurePath.is_reserved()` has changed in Python 3.13:

> _Changed in version 3.13_: Windows path names that contain a colon, or end with a dot or a space, are considered reserved. UNC paths may be reserved.

Besides the change in behavior, the `is_reserved()` method is deprecated as of Python 3.13 and will be removed in Python 3.15:

> _Deprecated since version 3.13, will be removed in version 3.15_: This method is deprecated; use [os.path.isreserved()](https://docs.python.org/3.13/library/os.path.html#os.path.isreserved) to detect reserved paths on Windows.

The change in behavior leads to `test_assert_valid_name_minimal` failing when asserting validity of path `"\n"`:

```python
=================================== FAILURES ===================================
________________________ test_assert_valid_name_minimal ________________________
setup_teardown_folder = (PosixPath('/tmp/pytest-of-mockbuild/pytest-0/test_assert_valid_name_minimal0'), PosixPath('/tmp/pytest-of-mockbuild/p...d_name_minimal0/test.exdir'), PosixPath('/tmp/pytest-of-mockbuild/pytest-0/test_assert_valid_name_minimal0/exdir_dir'))
    def test_assert_valid_name_minimal(setup_teardown_folder):
        f = exdir.File(setup_teardown_folder[1], name_validation=fv.minimal)
        exob._assert_valid_name("abcdefghijklmnopqrstuvwxyz1234567890_-", f)
        with pytest.raises(NameError):
            exob._assert_valid_name("", f)
    
        exob._assert_valid_name("A", f)
    
>       exob._assert_valid_name("\n", f)
tests/test_help_functions.py:36: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../BUILDROOT/python-exdir-0.5.0.1-5.fc41.x86_64/usr/lib/python3.13/site-packages/exdir/core/exdir_object.py:22: in _assert_valid_name
    container.file.name_validation(container.directory, name)
../../BUILDROOT/python-exdir-0.5.0.1-5.fc41.x86_64/usr/lib/python3.13/site-packages/exdir/core/validation.py:86: in minimal
    _assert_nonreserved(name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
name = '\n'
    def _assert_nonreserved(name):
        # NOTE ignore unicode errors, they are not reserved
        try:
            name_str = str(name)
        except UnicodeEncodeError:
            name_str = name.encode('utf8')
    
        reserved_names = [
            exob.META_FILENAME,
            exob.ATTRIBUTES_FILENAME,
            exob.RAW_FOLDER_NAME
        ]
    
        if name_str in reserved_names:
            raise NameError(
                "Name cannot be '{}' because it is a reserved filename in Exdir.".format(name_str)
            )
    
        if pathlib.PureWindowsPath(name_str).is_reserved():
>           raise NameError(
                "Name cannot be '{}' because it is a reserved filename in Windows.".format(name_str)
            )
E           NameError: Name cannot be '
E           ' because it is a reserved filename in Windows.
../../BUILDROOT/python-exdir-0.5.0.1-5.fc41.x86_64/usr/lib/python3.13/site-packages/exdir/core/validation.py:62: NameError
```

--- a/exdir/core/validation.py
+++ b/exdir/core/validation.py
@@ -1,13 +1,24 @@
 from enum import Enum
 import os
-try:
-    import pathlib
-except ImportError as e:
-    try:
-        import pathlib2 as pathlib
-    except ImportError:
-        raise e
+import sys
+from pathlib import Path, WindowsPath
+from unicodedata import category
 from . import constants as exob
+
+# `ntpath.isreserved` forbids ASCII control characters
+# (https://github.com/python/cpython/blob/7c016deae62308dd1b4e2767fc6abf04857c7843/Lib/ntpath.py#L325)
+# while `pathlib.PureWindowsPath.is_reserved` does not, so it is easiest to
+# forbid all control characters.
+if sys.version_info.minor < 13:
+    from pathlib import PureWindowsPath
+
+    def _is_reserved(path):
+        return PureWindowsPath(path).is_reserved() or _contains_control_character(path)
+else:
+    from ntpath import isreserved
+
+    def _is_reserved(path):
+        return isreserved(path) or _contains_control_character(path)
 
 VALID_CHARACTERS = ("abcdefghijklmnopqrstuvwxyz1234567890_-.")
 
@@ -17,6 +28,11 @@ class NamingRule(Enum):
     STRICT = 2
     THOROUGH = 3
     NONE = 4
+
+
+def _contains_control_character(s):
+    return any(ch for ch in s if category(ch)[0] == "C")
+
 
 def _assert_unique(parent_path, name):
     try:
@@ -58,7 +74,7 @@ def _assert_nonreserved(name):
             "Name cannot be '{}' because it is a reserved filename in Exdir.".format(name_str)
         )
 
-    if pathlib.PureWindowsPath(name_str).is_reserved():
+    if _is_reserved(name_str):
         raise NameError(
             "Name cannot be '{}' because it is a reserved filename in Windows.".format(name_str)
         )
@@ -102,7 +118,7 @@ def thorough(parent_path, name):
     name_lower = name_str.lower()
     _assert_valid_characters(name_lower)
 
-    if isinstance(pathlib.Path(parent_path), pathlib.WindowsPath):
+    if isinstance(Path(parent_path), WindowsPath):
         # use _assert_unique if we're already on Windows, because it is much faster
         # than the test below
         _assert_unique(parent_path, name)

--- a/tests/test_help_functions.py
+++ b/tests/test_help_functions.py
@@ -33,7 +33,8 @@ def test_assert_valid_name_minimal(setup_teardown_folder):
 
     exob._assert_valid_name("A", f)
 
-    exob._assert_valid_name("\n", f)
+    with pytest.raises(NameError):
+        exob._assert_valid_name("\n", f)
 
     exob._assert_valid_name(six.unichr(0x4500), f)
 

